Add ZTF/PS1 cross-match notebook

zclaytor opened the pull request:
This notebook runs some benchmark testing for the performance of LSDB cross-matches on Fornax with and without Dask parallelization. 

There are to-dos before this can be merged, but I figured better to get this started than to wait.

To-do:
- I made this notebook before the LSDB update/switch to HATS, so it's no longer able to access the catalogs. I'll try to figure out what the URIs for the HATS format catalogs.
- Fix a permission error that wasn't there before. This may or may not be related to the hipscat --> hats conversion

jkrick commented:
Hi Zach, thanks for contributing a notebook!
I just pushed a PR using panstarrs hats for a different notebook, and it is a pretty easy transition from lsdb.read_hipscat to lsdb.read_hats.  feel free to check that out for a template.  The new code is here: https://github.com/nasa-fornax/fornax-demo-notebooks/blob/main/light_curves/code_src/panstarrs_functions.py

bsipocz reviewed (commented):
Some first round comments

troyraen commented on crossmatch/ztf_ps1_crossmatch.md:
```diff
@@ -0,0 +1,158 @@
+cone_filter = ConeSearch(cone_ra, cone_dec, radius_arcsec)
+
+# Read ZTF DR20
+ztf_path = ("s3://irsa-mast-tike-spitzer-data/data/ZTF/dr20/objects/hipscat/ztf-dr20-objects-hipscat")
```
This HiPSCat version of ZTF DR20 is no longer available, but we should have a HATS version of ZTF DR23 available soon.

zclaytor replied:
Aha, I was wondering about this! Thanks for the info.

zclaytor commented on crossmatch/ztf_ps1_crossmatch.md:
```diff
@@ -0,0 +1,158 @@
+
+## Install LSDB
+
+Fornax has LSDB installed, but this notebook was written for lsdb v0.3.0, and it breaks with other versions. I haven't had time to figure out the specifics, so for now let's stick with v0.3.0.
```
Thank you! I've updated this to follow the example in the PS1 notebook linked above by specifying lsdb>=0.4.5 in the requirements file.

zclaytor commented on crossmatch/ztf_ps1_crossmatch.md:
````diff
@@ -0,0 +1,158 @@
+Fornax has LSDB installed, but this notebook was written for lsdb v0.3.0, and it breaks with other versions. I haven't had time to figure out the specifics, so for now let's stick with v0.3.0.
+
+```{code-cell} ipython3
+%pip install git+https://github.com/astronomy-commons/lsdb.git@v0.3.0
````
I appreciate the link to the guidelines. I've updated the notebook to better conform to the guidelines.

zclaytor commented:
I've addressed the comments above, and thank you all for the feedback! I believe the only remaining issue is the ZTF catalog link, which will be superseded by the HATS ZTF DR23 catalog when it becomes available.

jkrick commented:
Hi Zach, thanks for making those changes. 
There are two more sets of checklists that we use for notebooks going into this repo, a science review checklist and a tech review checklist (https://github.com/nasa-fornax/fornax-documentation/blob/62db23cd887ccd310d89cf8585632e849081da43/notebook_review_process.md).  Generally the process is that when you think the notebook is ready, you trigger a science review and a tech review of the notebook by others on the team.  I can do the science review, not sure if you have recommendations for someone to do the tech review.

I know this PR is waiting for ZTF HATS. That dataset will probably be ready in 2 - 4 weeks.  After that just let me know if you would prefer to a) merge this PR and make a future PR for the science/tech review stuff, or b) put everything into this PR.  I personally prefer to declare victory on a first PR without it needing to be perfect in every way, but it is up to you.

jrhastro commented:
> I know this PR is waiting for ZTF HATS. That dataset will probably be ready in 2 - 4 weeks. After that just let me know if you would prefer to a) merge this PR and make a future PR for the science/tech review stuff, or b) put everything into this PR. I personally prefer to declare victory on a first PR without it needing to be perfect in every way, but it is up to you.

@jkrick @troyraen Is there a GitHub issue or somewhere we can keep updated on the ZTF HATS release?

troyraen commented:
@jrhastro I think all of the existing tickets for that are internal to IRSA but I will open one in this repo for it soon. I'm currently adjusting the files according to the recent HATS format changes. I hope to have the datasets available in a testing bucket that is accessible from the Fornax Science Console around the middle of next week. I'll give you those paths when I have them.

I don't have a good timeline right now for when they will be released in the public bucket. If that doesn't happen by the time this PR is ready to merge then you can leave your notebook connected to the testing bucket and open an issue to switch the notebook to use the public bucket when it's available.

Also, in case you're interested in the Euclid Q1 MER Catalogs in HATS format, I recently put IRSA's version in that same testing bucket. I will post the info in a new issue.

troyraen commented:
> Also, in case you're interested in the Euclid Q1 MER Catalogs in HATS format, I recently put IRSA's version in that same testing bucket. I will post the info in a new issue.

https://github.com/nasa-fornax/fornax-demo-notebooks/issues/394

troyraen commented:
HATS version of ZTF DR23 Objects Table is now in the S3 testing bucket (available from Fornax only), @zclaytor @jrhastro. You will need `lsdb>=0.5.0` in order to read them. In addition, you'll need `UPath` for the testing bucket. Let me know if you run into any trouble.

```python
import lsdb
# Need UPath for the testing bucket. Otherwise hats will ignore the credentials that Fornax
# provides under the hood. Will be unnecessary after the dataset is released in a public bucket.
from upath import UPath

# Read with lsdb.
ztf_lsdb = lsdb.read_hats(UPath('s3://irsa-fornax-testdata/ZTF/dr23/objects/hats/'))
```

Once the Lightcurves are there as well, I'll open an issue with the path info similar to what I did for Euclid.

jrhastro commented:
@troyraen Thanks so much for generating these and for letting us know they are available!

zclaytor commented:
Thanks for the clarification! In that case, I believe it is ready for technical review.

vandesai1 commented:
The notebooks in this repo need to serve two purposes: 1) provide "realistic" use cases to clarify and test the resource allocation required to meet the goals of Fornax; 2) show users how to carry out tasks in the cloud that might be different from how they would do things on-prem. This notebook satisfies (2) so let's proceed with the other aspects of the review.

troyraen reviewed (approved):
Thanks @zclaytor! It's great to see some work on accessing these HATS catalogs using lsdb. This is a new catalog format and python library, and we'll definitely need to provide guidance to Fornax users through these these notebooks about how to use lsdb + dask efficiently for common tasks like cross matching.

I reviewed your notebook based on the tech-review checklist, which you can see on the [How Does Fornax Review Notebooks?](https://github.com/nasa-fornax/fornax-documentation/blob/main/notebook_review_process.md) page. For this first PR[*], it looks good to me in terms of meeting the baseline requirements to submit the notebook to the repo. There are four things that I do recommend (not require) that you consider for this PR:

- (1) Explicitly close the client and the cluster.
- (2) Show your benchmark results in a figure rather than a table.
- Revise the languaging around (3) "multiple nodes", and (4) "no dask".

I left a comment below about each of them.

Looking forward, I've also left comments below related to some of the other items on both checklists to give you some ideas to think about for if/when you want to start the next step to have your notebook published on the website (here: https://nasa-fornax.github.io/fornax-demo-notebooks/). In particular, two things about lsdb + dask that I've been thinking we'll need to cover in these Fornax notebooks are exactly what you've started to experiment with here -- how to choose the number of dask workers and how to manage the workers' memory usage effectively. If you want to do more revising on the code and narrative in the future, I think you could come up with some clear conclusions and recommendations that Fornax users can apply to their own science use cases and so would be especially valuable to have here. Two things that I think could help in terms of being able to make those kinds of conclusions/recommendations are to sort out the dask memory issues and to connect your notebook to a science use case. If the memory issues are what I think they are, I'm guessing that the workers can do more with the same amount of total resources (CPU and RAM) if you help dask manage the memory a little better, and I left some suggestions below that you might want to try. Having the notebooks motivated by a science use case is good for a few reasons, but I bring it up here specifically because we might find that it changes the results in significant ways, and I'll say more about that below as well.

[*] I see that we haven't really described this bit in the Review docs yet, but notebooks that go through the full process to get published on the website tend to do it in two-ish steps -- first is the initial submission to the repo that goes through a baseline review, and then the second is another 1+ PRs to iterate on the notebook content and hook it in to the testing/CI/publishing workflows. You can see, for example, that notebooks under the forced_photometry and light_curves (aka, "time domain") use cases/directories have been published, but neither of the spectroscopy notebooks have made it all the way to the website yet. In the second step, we do more thorough reviews and iterations of the notebook content to cover the full science- and tech-review checklists before publishing. We'll have more group discussions after everyone is back in order to reach consensus about how to better define and formalize these aspects of the review process. But I don't think you need to wait for all of that to be worked out before merging this PR. (But do wait at least for @jkrick's review.)

jkrick commented on crossmatch/ztf_ps1_crossmatch.md:
```diff
@@ -0,0 +1,192 @@
+    client
+
+# Select the search radius to give us the right number of rows
+radius = { # Nrows: radius_arcseconds
```
please add a sentence about how you generated this radius dict.

jkrick commented on crossmatch/ztf_ps1_crossmatch.md:
```diff
@@ -0,0 +1,192 @@
+For each level, we want to know the performance with (1) no Dask, (2) minimal Dask - like 2 workers, (3) bigger Dask - as many workers as we can use, and (4) auto-scaling Dask.
+
++++
+
```
Please add a "Runtime" section to the intro.  You can use this notebook for an example of what that should look like: https://github.com/nasa-fornax/fornax-demo-notebooks/blob/main/light_curves/light_curve_generator.md

or the tech review checklist under "documentation" has an example.

jkrick reviewed (approved):
Hi Zach, I have added comments below for the science review.  I have approved the PR so I don't need to look at it again after you make changes.

troyraen reviewed (approved):
Thanks for doing so much work here, @zclaytor.

One thing to change throughout: Somewhere in the Introduction, explain that as of August 2025 the Fornax Science Console server types were w, x, y, z (fill in names and specs) and that is why the specific number of CPUs and RAM were chosen for these benchmarks. Then, in everything below the Runtime section, please remove all references to the server types and just use the specs (eg, number of CPUs) instead. This will make the notebook more understandable to everyone. Also, the fornax server types can and do change periodically, so this will make the notebook more future proof.

One FYI: I image that we'll end up putting this in a different directory that is shared with other notebooks and then update the readme, etc. to match. But we'll sort that out later. Nothing for you to do about it now.

Specific comments below.

troyraen commented:
Hi @zclaytor, just wanted to let you know that I'm the only one reviewing this round, so you don't need to wait for reviews from anyone else here. Also, not all of the suggestions above need to be taken care of in this PR. The two things that would be great for you to update before merging, if you can, are:

- Add a comment about why the upper-limit pin `numpy<=2.2` is needed.
- Update text in the sections "Record and plot benchmarks" and "Summary" to point out what seems to work best for larger sample sizes since those are central to `lsdb` use cases.

Let me know if you have questions/concerns about those. I've already approved, so you can merge after that when you're ready. I feel that the other changes would be helpful before completing the second step to release the notebook publicly but they can wait to be discussed and/or implemented in a future PR if you don't want deal with them here. You and/or someone else working in this repo can do future PRs.

bsipocz commented on crossmatch/xmatch_benchmarks.csv:
Do we need to get this file in the repo? if yes, consider putting it either the data or the output directory, like we do for the other notebooks

bsipocz reviewed (commented):
I haven't tried running this notebook, but from just looking at it my only comment is about the data file.
1) do we need it in the repo?
2) if yes, we should put it to a more standardised place, e.g. in an `output` directory.

troyraen commented:
Thanks so much for working through all of that @zclaytor. Really awesome that you were able do the full-catalog cross match! Will be great to have this notebook in. I tested it on the fornax console and it ran successfully.

troyraen commented:
@zclaytor @jrhastro this is deployed and rendered on the site now! https://nasa-fornax.github.io/fornax-demo-notebooks/crossmatch/ztf_ps1_crossmatch.html
